Screen Border Fade-Out

ulfbiallas opened the pull request:
**Changes**

This PR adds a new parameter `ssr_screen_border_fadeout`.
It can be used to control a fade-out effect of reflections close to the screen borders
and supports a range from `0` (fade-out disabled) to `1` (max fade-out). The default value is `0`.


If you move the camera so that a reflected object is no longer visible,
its reflection on the water also disappears, even though it should still be visible.
This is a general weakness of SSR, as only objects that are visible can be reflected.
Enabling screen border fade-out mitigates this to some degree and helps to make it less distracting.


This implementation is based on https://github.com/marcelb/GodotSSRWater/pull/3 by @Xtarsia and another variant, which was proposed in https://github.com/marcelb/GodotSSRWater/issues/2 by @farfalk.

I decided to change a couple of things to make it easier to use and also the code easier to understand:
- Setting the parameter to `0` disables the effect
- The logic is moved to its own function.
- Calculate the transparency simply based on distance from the screen center.

Xtarsia committed:
Update water.gdshader Add SSR edge smoothing

Also added Edge Sharpness uniform

ulfbiallas committed:
Merge branch 'main' into edge-smoothing

ulfbiallas committed:
refactor code

ulfbiallas committed:
adjust smoothing calculation

ulfbiallas committed:
rename parameter

marcelb reviewed (changes requested):
We have three different implementations for edge fading now!

I created all three masks in shadertoy (https://www.shadertoy.com/new):

@farfalk
![chrome_2vdsdJrysP](https://github.com/marcelb/GodotSSRWater/assets/308374/d0d6b964-30ce-475a-9969-c7be5bea8ec1)

```
void mainImage( out vec4 fragColor, in vec2 fragCoord )
{
    // Normalized pixel coordinates (from 0 to 1)
    vec2 uv = fragCoord/iResolution.xy;

    vec2 centered_screen_coords = smoothstep(0.2, 0.6, abs(vec2(0.5, 0.5) - uv.xy));

    float screen_edge_factor = clamp(1.0 - (centered_screen_coords.x + centered_screen_coords.y), 0.0, 1.0);

    fragColor = vec4(screen_edge_factor, 0.0, 0.0, 1.0);
}
```

@ulfbiallas 
![chrome_35XMOvvmb2](https://github.com/marcelb/GodotSSRWater/assets/308374/5f20f9a8-f397-436f-8e81-9525c846bd13)

```
void mainImage( out vec4 fragColor, in vec2 fragCoord )
{
    // Normalized pixel coordinates (from 0 to 1)
    vec2 uv = fragCoord/iResolution.xy;

    vec2 screen_pos = ((2.0 * uv) - 1.0);

    float ssr_screen_border_fadeout = 0.8;
    float col = clamp(1.0 - (screen_pos.x * screen_pos.x + screen_pos.y * screen_pos.y) * ssr_screen_border_fadeout, 0.0, 1.0);

    fragColor = vec4(col, 0.0, 0.0, 1.0);
}
```

@Xtarsia 
![chrome_GE7BYeMAr2](https://github.com/marcelb/GodotSSRWater/assets/308374/cd45e169-6cd7-4dfb-8e04-552de173abc1)

```
void mainImage( out vec4 fragColor, in vec2 fragCoord )
{
    // Normalized pixel coordinates (from 0 to 1)
    vec2 uv = fragCoord/iResolution.xy;

    uv *= 1.0 - uv;

    float edgeSharpness = 0.5;
    float col = clamp(uv.x * uv.y * 16.0 * (1.0 + edgeSharpness), 0.0, 1.0);

    fragColor = vec4(col, 0.0, 0.0, 1.0);
}
```

I think the mask itself that makes the most sense is the one by @Xtarsia right now. It keeps the rectangle shape of the screen and loses the least viable pixels, while keeping always a black border (full fade out), ensuring smooth edges in all cases. The downside is, that you can't switch it off, you can just put the edge sharpness to really high values, so it stops being visible.

@farfalks version is also pretty good, but lacks a value to customize it and comes with the problem, that the top and side edges are smoothed out, but will be non-black, so a possible ssr edge can still be visible and wont be smoothed out.

@ulfbiallas version has the smoothest fadeout, but I think it also has the problem, that the edges are not really faded out (black in shader toy) on lower values. That leads to this:

![Godot_v4 2 2-stable_win64_w2sfj6qn1E](https://github.com/marcelb/GodotSSRWater/assets/308374/958ac477-be34-4c2b-bf09-5f3c2a0fc037)

Even though it starts fading out, the hard edge is still visible.

It's because the mask edges wont be really black:
![chrome_35XMOvvmb2__](https://github.com/marcelb/GodotSSRWater/assets/308374/2d5c195b-f708-4eb9-9b1a-e1d8a15f0be6)

Only with high values of fadeout (like 1.0) its actually black (but this throws away a lot of pixels that could have been displayed for SSR):
![chrome_VbSYvf9UWQ](https://github.com/marcelb/GodotSSRWater/assets/308374/32fad14c-8f46-458d-9413-20a07f53747c)

Only @Xtarsia mask keeps that on most values (except really high ones):
![chrome_GE7BYeMAr2__](https://github.com/marcelb/GodotSSRWater/assets/308374/4cb54437-f824-46e4-b0df-269c4412f43e)

I think the best version would be a a squarish mask that is black on the edges, so its never possible to see a hard edge, which uses a fadeout_strength configuration instead of a sharpness_strength configuration and can also be deactivated.

Hope that makes sense.

marcelb reviewed (commented):
We also have to add a new configuration field to the README.md :)

Xtarsia commented:
```
void mainImage( out vec4 fragColor, in vec2 fragCoord )
{
    // Normalized pixel coordinates (from 0 to 1)
    vec2 uv = fragCoord/iResolution.xy;

    uv *= 1.0 - uv;

    float edgeFade = 2.0; // Range 0.0 (Fade disabled) to 8.0~ (Fully faded out..)
    float col = smoothstep(0.0, edgeFade, uv.x * uv.y * 64.0);

    fragColor = vec4(col, 0.0, 0.0, 1.0);
}
```

smoothstep version, a tad more square and the fade is disabled with a value of 0.0

marcelb commented:
0.5
![chrome_zDuuKnoNjq](https://github.com/marcelb/GodotSSRWater/assets/308374/243db0c5-e738-46a1-8dd7-5e54e01f05aa)

2.0
![chrome_5UD0JMqOmw](https://github.com/marcelb/GodotSSRWater/assets/308374/83b7f9df-ee26-492b-aae2-70520c99a8a7)

0.0
![chrome_EulrKhl19Y](https://github.com/marcelb/GodotSSRWater/assets/308374/ad0fe97a-76fd-4801-8887-e51d1125724e)

Looks amazing. @Xtarsia

ulfbiallas committed:
Merge branch 'main' into edge-smoothing

ulfbiallas committed:
change mask

ulfbiallas commented:
Yes, that looks much better! I see the advantage of this variant: The fading behaves more similar at the corners as well as at the edges of the screen. I updated this PR to implement this with some adjustments: The parameter now ranges exactly from 0 to 1.

Here is the shadertoy code to try yourself. I mapped the parameter to the x-position of the mouse.
You can play with it by dragging the mouse over the canvas.

```
void mainImage( out vec4 fragColor, in vec2 fragCoord ) {

    vec2 uv = fragCoord/iResolution.xy;
    
    float edgeFade = iMouse.x / iResolution.x;
    
    vec2 shiftedUv = 4.0 * uv * (1.0 - uv);
    float factor = shiftedUv.x*shiftedUv.y;
    float offset = mix(0.0, 0.5, (clamp(edgeFade, 0.75, 1.0) - 0.75) / 0.25);

    float alpha = smoothstep(0.0, 2.0*edgeFade, factor) - offset;

    fragColor = vec4(alpha, 0.0, 0.0, 1.0);
}
```

marcelb commented on shaders/water.gdshader:
```diff
@@ -84,7 +85,22 @@ vec3 get_view_position_from_uv(vec2 uv, float depth, mat4 inv_proj_m)
+	// This ensure alpha smoothly transitions to zero when ssr_screen_border_fadeout is approaching 1.0.	
+	float offset = mix(0.0, 0.5, (clamp(ssr_screen_border_fadeout, 0.75, 1.0)-0.75) / 0.25);
+ 
+	float alpha = smoothstep(0.0, 2.0 * ssr_screen_border_fadeout, mask) - offset;
```
```suggestion
	float alpha = clamp(smoothstep(0.0, 2.0 * ssr_screen_border_fadeout, mask) - offset, 0.0, 1.0);
```

here is a clamp missing. Otherwise negative alphas can happen (starts with fadeout 0.8+):
![Godot_v4 2 2-stable_win64_gyvmeb8xgD](https://github.com/marcelb/GodotSSRWater/assets/308374/f8b779f8-5fc1-4c10-a6d5-c5f90401c34c)

Example of static -0.5 Alpha value to illustrate:
![Godot_v4 2 2-stable_win64_pNnlYbJkAW](https://github.com/marcelb/GodotSSRWater/assets/308374/23c0201e-83d9-42c6-a7bc-f27b23c01c16)

ulfbiallas replied:
Good catch! will add it

marcelb commented on World.tscn:
```diff
@@ -87,6 +87,7 @@ shader_parameter/ssr_resolution = 2.0
 shader_parameter/ssr_max_travel = 30.0
 shader_parameter/ssr_max_diff = 4.0
 shader_parameter/ssr_mix_strength = 0.7
+shader_parameter/ssr_screen_border_fadeout = 0.0
```
I'd prefer 0.3 standard value, since it has great visual value.

```suggestion
shader_parameter/ssr_screen_border_fadeout = 0.3
```

marcelb reviewed (changes requested):
I think we should just fix those last issues and add the new parameter description to README.md and go with it!

marcelb commented on shaders/water.gdshader:
```diff
@@ -19,6 +19,7 @@ uniform float ssr_resolution   : hint_range(0.01, 10.0, 0.1)	= 2.0;
 uniform float ssr_max_travel   : hint_range(0.0, 200.0, 0.1) 	= 30.0;
 uniform float ssr_max_diff     : hint_range(0.1, 10.0, 0.1) 	= 4.0;
 uniform float ssr_mix_strength : hint_range(0.0, 1.0, 0.01) 	= 0.7;
+uniform float ssr_screen_border_fadeout: hint_range(0.0, 1.0, 0.1)	= 0.0;
```
```suggestion
uniform float ssr_screen_border_fadeout: hint_range(0.0, 1.0, 0.1)	= 0.3;
```

Commit:
Enable screen border fade-out by default

Co-authored-by: Marcel Bankmann <adsci81@gmail.com>

ulfbiallas committed:
add missing clamp

ulfbiallas committed:
update readme

ulfbiallas committed:
skip calculation when param is 0

ulfbiallas commented:
@marcelb I just updated the branch with your feedback and also added a check to skip the calculation in the case when `ssr_screen_border_fadeout == 0.0`

marcelb committed:
Update World.tscn

marcelb commented:
Thank you everyone @ulfbiallas @ulfbiallas @farfalk, I merged this!
